### DISCUSSION TITLE
--log-file/--stdout/--stderr options ignored

### DISCUSSION THREAD

**Discussion by Gollum999** [apache/airflow#21843] [General]
### Apache Airflow version

2.2.3 (latest released)

### What happened

Some airflow processes (I've tested `webserver`, `scheduler`, `triggerer`, `celery worker`, and `celery flower`) ignore the `--log-file`, `--stdout`, and `--stderr` options when running without `--daemon`.  No log files are created, and logs continue being printed to stdout/stderr.  This can be a bit confusing in environments where daemonization is handled externally, e.g. when running through [systemd](https://airflow.apache.org/docs/apache-airflow/stable/howto/run-with-systemd.html).

When `--daemon` is added, most process types will suppress stdout/stderr and write to the specified log files as expected.  **However, `webserver` does not write any logs except for a few lines on shutdown, and since stdout/stderr are suppressed, all useful logs are lost.**  (Though worth noting that `--access-logfile` and `--error-logfile` work as expected, with or without `--daemon`.)

### What you expected to happen

Command line args should be respected whenever possible.  If there are certain combinations of args that are incompatible, they should be noted as such in the `--help` and/or emit a warning at runtime.

With `airflow webserver` in particular, these arguments seem to be mostly useless since no (useful) logs are written to a file, regardless of whether or not `--daemon` is specified.  If the webserver behavior cannot be changed, maybe these options should just be removed.

### How to reproduce

With a fresh install (no changes to `airflow.cfg`, etc):

```
airflow webserver --log-file=$AIRFLOW_HOME/webserver.log --stdout=$AIRFLOW_HOME/webserver.out --stderr=$AIRFLOW_HOME/webserver.err  # no logs created, stdout/stderr not redirected
ls -la $AIRFLOW_HOME/webserver.*
airflow webserver --log-file=$AIRFLOW_HOME/webserver.log --stdout=$AIRFLOW_HOME/webserver.out --stderr=$AIRFLOW_HOME/webserver.err --daemon  # logs created but remain empty until webserver is killed, stdout/stderr suppressed after startup
kill $(cat $AIRFLOW_HOME/airflow-webserver-monitor.pid)
ls -la $AIRFLOW_HOME/webserver.*

airflow scheduler --log-file=$AIRFLOW_HOME/scheduler.log --stdout=$AIRFLOW_HOME/scheduler.out --stderr=$AIRFLOW_HOME/scheduler.err  # no logs created, stdout/stderr not redirected
ls -la $AIRFLOW_HOME/scheduler.*
airflow scheduler --log-file=$AIRFLOW_HOME/scheduler.log --stdout=$AIRFLOW_HOME/scheduler.out --stderr=$AIRFLOW_HOME/scheduler.err --daemon  # seems to work as expected
kill $(cat $AIRFLOW_HOME/airflow-scheduler.pid)
ls -la $AIRFLOW_HOME/scheduler.*

# Repeat as necessary for other process types
```

### Operating System

CentOS 7.4

### Versions of Apache Airflow Providers

N/A

### Deployment

Other

### Deployment details

Just running locally, no containers.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
I think it would be great to update the help on that. PR to clarify the behaviours are most welcome. Would you like to update the help as appropriate? You can easily become one of the almost 2000 contributors to airflow ? This seems like a cool contribution to review and update the messages related to those flags. I think, since you just experienced it, you are one of the best people to do it @Gollum999. The PRs for that should be rather easy.

The right descriptions/help messages can be found here https://github.com/apache/airflow/tree/main/airflow/cli/commands

**Comment by kadai0308** [comment]
Hi @potiuk, nice to meet you. I am Kadai, new here.

For the webserver part, after go through the code, for writing the log when daemon is off,  is it just write the output to a file at [here](https://github.com/apache/airflow/blob/main/airflow/cli/commands/webserver_command.py#L475) ?

something like:
```
with subprocess.Popen(run_args, close_fds=True, stdout=out_logfile, stderr=error_logfile) as gunicorn_master_proc:
```
Not sure am i missing anything else need to modify, if not, then i think i can help on this issue.

**Comment by potiuk** [comment]
I am not sure if understand. I will convert it into discussion and maybe the @Gollum999 and @kadai0308 you can can make a proposal what to do with it.

**Comment by 1paridis** [comment]
maybe you can use logrotate to do this. linux has logrotate to rotate the log. for example.
when you run airflow webserver, you can use this `nohup airflow webserver --port 8080 >> ./airflow/web.log 2>&1 & `. then you have a server log file `web.log`。 
Now do it like this.
## touch a logrotate config file
```bash
cat << EOF >>./airflow_webserver
/data/airflow/web.log {
    daily
    rotate 7
    missingok
    dateext
    copytruncate
}
EOF 
```
> notice the web.log path.

## run test

```bash
logrotate -vf ./airflow_webserver
```

you will see some log on the console.

## define a job 
you can use crontab or airflow to schedule the job .
about the logrotate you can google it.

**Comment by bruckwubete** [comment]
@Gollum999 It does work. You must run it as a daemon by adding a -D flag! I have verified this in v2.7.1
